pandas/io/parsers.py

```diff
@@ -801,6 +803,27 @@ def _get_options_with_defaults(self, engine):
 
         return options
 
+    def _check_file_or_buffer(self, f, engine):
+        # see gh-16530
+        if is_file_like(f):
+            next_attr = "__next__" if PY3 else "next"
+
+            # The C engine doesn't need the file-like to have
+            # the "__next__" attribute. However, the Python
+            # explicitly calls "next(...)" when iterating through
+            # such an object, meaning it needs to have that attribute.
+            if engine != "c" and not hasattr(f, next_attr):
+                msg = ("The 'python' engine can not iterate "
```

Comment:
this block seems untested. Are you able to add some, or is it a hassle to hit it?